File: 3.Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        maxlen = 0
        sett = set()
        l = 0

        for r in range(len(s)):
            while s[r] in sett:
                sett.remove(s[l])
                l += 1
            sett.add(s[r])
            maxlen = max(maxlen, r-l+1)
        return maxlen
            
        """
        :type s: str
        :rtype: int
        """

File: 3.Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxlen = 0; 
        std::set <char> letter_set; //set 
        int l = 0; //initialize left pointer 

        for(int r = 0; r < s.length(); r++)
        {
            while(letter_set.contains(s[r])) { //if character in set already
                letter_set.erase(s[l]); //shift left pointer and remove char from set 
                l++; 
            }
            letter_set.insert(s[r]); //new letter that is part of substring 
            maxlen = max(maxlen, r-l+1); //solution is max of current length substring and max 
        }
        return maxlen;
    }
};